src/common/lua/ngx_wasm_lua_ffi.c

```diff
@@ -304,126 +288,126 @@ ngx_wa_ffi_shm_setup_zones(ngx_wa_ffi_shm_setup_zones_handler_pt setup_handler)
+    ngx_str_t **keys)
 {
-    ngx_wa_shm_kv_node_t *n = (ngx_wa_shm_kv_node_t *) node;
+    ngx_wa_shm_kv_node_t  *n = (ngx_wa_shm_kv_node_t *) node;
 
-    if (limit > 0 && (*total - start) == limit) {
+    if ((*cur_idx - start_idx) == limit) {
         return;
     }
 
-    if (keys && *total >= start) {
-        keys[(*total - start)] = &n->key.str;
+    if (*cur_idx >= start_idx) {
+        /* append to result */
+        keys[(*cur_idx - start_idx)] = &n->key.str;
     }
 
-    (*total)++;
+    (*cur_idx)++;
 
     if (node->left != sentinel) {
-        shm_kv_retrieve_keys(node->left, sentinel, start, limit, keys, total);
+        retrieve_keys(node->left, sentinel, limit, start_idx, cur_idx, keys);
     }
 
     if (node->right != sentinel) {
-        shm_kv_retrieve_keys(node->right, sentinel, start, limit, keys, total);
+        retrieve_keys(node->right, sentinel, limit, start_idx, cur_idx, keys);
     }
 }
 
 
 ngx_int_t
-ngx_wa_ffi_shm_iterate_keys(ngx_wa_shm_t *shm,
-                            ngx_uint_t page, ngx_uint_t page_size,
-                            ngx_str_t **keys, ngx_uint_t *total)
+ngx_wa_ffi_shm_iterate_keys(ngx_wa_shm_t *shm, ngx_uint_t page_size,
+    ngx_uint_t *clast_idx, ngx_uint_t *cur_idx, ngx_str_t **keys)
 {
+    ngx_uint_t        last_idx = clast_idx ? *clast_idx : 0;
     ngx_wa_shm_kv_t  *kv;
 
-    ngx_wa_assert(shm && shm->type != NGX_WA_SHM_TYPE_QUEUE);
+    ngx_wa_assert(shm->type != NGX_WA_SHM_TYPE_QUEUE);
 
     if (!ngx_wa_shm_locked(shm)) {
         return NGX_ABORT;
     }
 
     kv = shm->data;
 
-    if (page >= kv->nelts) {
-        return NGX_DECLINED;
+    if (last_idx >= kv->nelts) {
+        /* finished */
+        return NGX_DONE;
     }
 
-    if (kv->rbtree.root != kv->rbtree.sentinel) {
-        shm_kv_retrieve_keys(kv->rbtree.root, kv->rbtree.sentinel,
-                             page, page_size, keys, total);
-    }
+    /* cannot be empty here (kv->nelts > 0) */
+    ngx_wa_assert(kv->rbtree.root != kv->rbtree.sentinel);
 
-    *total = *total - page;
+    retrieve_keys(kv->rbtree.root, kv->rbtree.sentinel, page_size, last_idx,
+                  cur_idx, keys);
+
+    *cur_idx -= last_idx;
+
+    if (clast_idx) {
+        *clast_idx += *cur_idx;
+    }
 
     return NGX_OK;
 }
 
 
+ngx_uint_t
+ngx_wa_ffi_shm_kv_nelts(ngx_wa_shm_t *shm)
+{
+    ngx_wa_shm_kv_t  *kv;
+
+    ngx_wa_assert(shm->type != NGX_WA_SHM_TYPE_QUEUE);
+
+    kv = shm->data;
+
+    return kv->nelts;
+}
+
+
 ngx_int_t
-ngx_wa_ffi_shm_get_kv_value(ngx_wa_shm_t *shm,
-                            ngx_str_t *k, ngx_str_t **v, uint32_t *cas)
+ngx_wa_ffi_shm_kv_get(ngx_wa_shm_t *shm, ngx_str_t *k, ngx_str_t **v,
+    uint32_t *cas)
 {
-    unsigned   unlock = 0;
     ngx_int_t  rc;
 
-    ngx_wa_assert(shm && k && v && cas);
+    ngx_wa_assert(shm->type == NGX_WA_SHM_TYPE_KV);
 
-    if (!ngx_wa_shm_locked(shm)) {
-        ngx_wa_shm_lock(shm);
-        unlock = 1;
-    }
+    ngx_wa_shm_lock(shm);
```

Comment:
This change prevents the user from being able to retrieve values while iterating over the keys of an `shm` 
```
shm.kv:lock()

for key in shm.kv:iterate_keys() do
    local value = shm.kv:get(key)
end

shm.kv:unlock()

Reply:
Fair - will also need tests.

Reply:
I included a test case for this situation in the original [PR](https://github.com/Kong/ngx_wasm_module/pull/582/files#diff-c41e29444a09e4c8bb9f78a60119c76724751ac81101d489497b794b758dfbe2R15-R53).

Do you want me to push a commit adding a new version of this test?

Reply:
I recall this test but the note was for myself. I must also say, I remembered this usage in the test but as I stated in our call, it was actually not properly isolated (testing multiple new APIs together in the same suite/test). A proper test case would have been separate from the sanity test and explicitly state the purpose of the test (e.g. "can call get() while iterating).

Reply:
I'm sorry, I misunderstood the note.

Thank you for clarifying the convention over test isolation and modularity. I wasn't mindful of those conventions while writing those tests, but will be next time.